core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * Simplified "Computation Reuse via Fusion in Amazon Athena" paper JoinOnKeys rule
+ * that works only on a single cross join over global aggregations over the same table.
+ * It transforms
+ * <pre>
+ *   cross join
+ *     aggregation global
+ *       agg_left <- aggLeft()
+ *         filter(filter_left)
+ *           table scan(table)
+ *     aggregation global
+ *       agg_right <- aggRight()
+ *         filter(filter_right)
+ *           table scan(table)
+ * </pre>
+ * into:
+ * <pre>
+ *   aggregation global
+ *       agg_left <- aggLeft(mask = agg_mask_left)
+ *       agg_right <- aggRight(mask = agg_mask_right)
+ *     project(agg_mask_left = filter_left, agg_mask_left = filter_right)
+ *       filter(filter_left or filter_right)
+ *         table scan
+ * </pre>
+ * <p>
+ * This makes the plan to read the source table only once,
+ * resulting in the potentially significant performance improvement.
+ */
+public class FuseCrossJoinedGlobalAggregations
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join()
+            .matching(JoinNode::isCrossJoin);
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isFuseSubPlanEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        PlanNode left = context.getLookup().resolve(node.getLeft());
+        PlanNode right = context.getLookup().resolve(node.getRight());
+
+        while (left instanceof ProjectNode && outputsSameAsSource((ProjectNode) left)) {
+            // ignore identity projection
+            left = context.getLookup().resolve(((ProjectNode) left).getSource());
+        }
+        while (right instanceof ProjectNode && outputsSameAsSource((ProjectNode) right)) {
+            // ignore identity projection
+            right = context.getLookup().resolve(((ProjectNode) right).getSource());
+        }
+
+        if (!(right instanceof AggregationNode) || !((AggregationNode) right).getGroupingKeys().isEmpty()) {
+            return Result.empty();
+        }
+
+        if (!(left instanceof AggregationNode) || !((AggregationNode) left).getGroupingKeys().isEmpty()) {
+            return Result.empty();
+        }
+
+        return new PlanNodeFuser(context).fuse(left, right)
+                .map(fused -> {
+                    PlanNode result;
+                    if (fused.leftFilter().equals(TRUE_LITERAL) && fused.rightFilter().equals(TRUE_LITERAL)) {
+                        result = fused.plan();
+                    }
+                    else {
+                        result = new FilterNode(
+                                context.getIdAllocator().getNextId(),
+                                fused.plan(),
+                                and(fused.leftFilter(), fused.rightFilter()));
```

Comment:
Isn't this have to be `or`?

Reply:
Not according to the paper

> JoinOnKeys
...
An interesting special case of
this rule happens when both K1 and K2 are empty (e.g., for
scalar aggregates) and the subsequent join is a cross product. In
this case, given GroupBy,A1(Q1)⨯GroupBy,A2(Q2) and
Fuse(Q1, Q2) = (Q, M, L, R), we transform the input
pattern into FilterL AND R(GroupBy,A1 M(A2)(Q)).

That said, i don't think I have it tested and I'm not even sure I can setup this case using current fuse. 
I will double check

Reply:
I think both left and right filters will always be `true` if you are not dealing with the non-scaler GroupBy

Reply:
yeah, currently this always gonna be true because fused aggregations do not expose filters (those are applied inside the aggregation using a mask)
I dropped the if and added precondition